tree/ntuple/v7/src/RNTupleIndex.cxx

```diff
@@ -115,31 +129,45 @@ void ROOT::Experimental::Internal::RNTupleIndex::Build()
 ROOT::Experimental::NTupleSize_t
 ROOT::Experimental::Internal::RNTupleIndex::GetFirstEntryNumber(const std::vector<void *> &valuePtrs) const
 {
-   const auto entryIndices = GetAllEntryNumbers(valuePtrs);
-   if (!entryIndices)
+   const auto entryNumbers = GetAllEntryNumbers(valuePtrs);
+   if (entryNumbers.empty())
       return kInvalidNTupleIndex;
-   return entryIndices->front();
+   return entryNumbers.front();
 }
 
-const std::vector<ROOT::Experimental::NTupleSize_t> *
+const std::vector<ROOT::Experimental::NTupleSize_t>
 ROOT::Experimental::Internal::RNTupleIndex::GetAllEntryNumbers(const std::vector<void *> &valuePtrs) const
 {
    if (valuePtrs.size() != fIndexFields.size())
       throw RException(R__FAIL("Number of value pointers must match number of indexed fields."));
 
    EnsureBuilt();
 
-   std::vector<NTupleIndexValue_t> indexValues;
-   indexValues.reserve(fIndexFields.size());
+   std::vector<std::vector<NTupleIndexValue_t>> entryNumbersPerCluster;
```

Comment:
This implementation uses two vectors (`entryNumbersPerCluster` and `entryNumbers`) which require two heap allocations. Would it be possible to fill `entryNumbers` directly?